File: cover/src/main/java/com/ruidoespontaneo/cassette/cover/components/WavyPill.kt
package com.ruidoespontaneo.cassette.cover.components

import androidx.compose.runtime.Composable
import androidx.compose.runtime.FloatState
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.withFrameNanos
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawWithCache
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import kotlin.math.PI
import kotlin.math.min
import kotlin.math.sin

/** Wave shape and speed for [wavyPillBackground], matching [PreviewWaveform]'s leading line. */
val WavyPillWavelength = 40.dp
val WavyPillWaveSpeed = 40.dp

/** How far [wavyPillBackground]'s edges swing at full amplitude — the pill is inset by this much. */
val WavyPillMaxAmplitude = 3.dp

/** Distance between the points a wavy edge is sampled at, in px. */
private const val SAMPLE_STEP_PX = 2f

/**
 * How far a wavy edge is displaced at [x], for a straight run from [start] to [end] (the pill's
 * flat stretch between its round caps). A sine of [wavelength], shifted by [phase], scaled by
 * [amplitude] — tapered to 0 over half a wavelength at either end so the wave meets the caps
 * without a kink. 0 outside the run.
 */
fun wavyEdgeOffset(x: Float, start: Float, end: Float, amplitude: Float, wavelength: Float, phase: Float): Float {
    if (x <= start || x >= end || amplitude == 0f) return 0f
    val taper = min(wavelength / 2f, (end - start) / 2f)
    val fromEdge = min(x - start, end - x)
    val envelope = if (fromEdge >= taper) 1f else smoothstep(fromEdge / taper)
    return amplitude * envelope * sin(2f * PI.toFloat() * (x - phase) / wavelength)
}

private fun smoothstep(t: Float): Float = t * t * (3f - 2f * t)

/**
 * Fills the bounds with a pill whose top and bottom edges are travelling waves, M3 Expressive's
 * wavy look (see [PreviewWaveform]) applied to a container. The pill is inset top and bottom by
 * [maxAmplitude] so the waves never leave the bounds; its ends stay round.
 *
 * [amplitude] is a 0..1 fraction of [maxAmplitude] (0 draws a plain pill) and [phase] how far the
 * wave has travelled, in px — both read at draw time, so animating them only redraws.
 */
fun Modifier.wavyPillBackground(
    color: Color,
    amplitude: () -> Float,
    phase: () -> Float,
    wavelength: Dp = WavyPillWavelength,
    maxAmplitude: Dp = WavyPillMaxAmplitude
): Modifier = drawWithCache {
    val path = Path()
    val inset = maxAmplitude.toPx()
    val wavelengthPx = wavelength.toPx()
    val bottom = size.height - inset
    val radius = ((bottom - inset) / 2f).coerceAtLeast(0f)
    val end = (size.width - radius).coerceAtLeast(radius)
    onDrawBehind {
        val swing = amplitude().coerceIn(0f, 1f) * inset
        val shift = phase()
        path.reset()
        // Top edge, left to right, then the right cap; bottom edge back, then the left cap. The
        // bottom mirrors the top, so the pill swells and pinches rather than bending.
        path.moveTo(radius, inset)
        var x = radius
        while (x < end) {
            x = min(x + SAMPLE_STEP_PX, end)
            path.lineTo(x, inset - wavyEdgeOffset(x, radius, end, swing, wavelengthPx, shift))
        }
        path.arcTo(Rect(end - radius, inset, end + radius, bottom), -90f, 180f, false)
        while (x > radius) {
            x = maxOf(x - SAMPLE_STEP_PX, radius)
            path.lineTo(x, bottom + wavyEdgeOffset(x, radius, end, swing, wavelengthPx, shift))
        }
        path.arcTo(Rect(radius - radius, inset, radius + radius, bottom), 90f, 180f, false)
        path.close()
        drawPath(path, color)
    }
}

/**
 * The travelling phase for [wavyPillBackground], in px. It advances at [speed] while [running] —
 * and, once that turns false, keeps going for as long as [isSettling] says the wave is still
 * visible, so it eases flat while moving instead of freezing mid-ripple. Idle, it asks for no frames.
 */
@Composable
fun rememberWavePhase(
    running: Boolean,
    isSettling: () -> Boolean,
    wavelength: Dp = WavyPillWavelength,
    speed: Dp = WavyPillWaveSpeed
): FloatState {
    val density = LocalDensity.current
    val wavelengthPx = with(density) { wavelength.toPx() }
    val speedPx = with(density) { speed.toPx() }
    val phase = remember { mutableFloatStateOf(0f) }
    LaunchedEffect(running, wavelengthPx, speedPx) {
        var last = withFrameNanos { it }
        while (running || isSettling()) {
            withFrameNanos { now ->
                phase.floatValue = (phase.floatValue + speedPx * (now - last) / 1_000_000_000f) % wavelengthPx
                last = now
            }
        }
    }
    return phase
}

File: iosApp/iosApp/Common/WavyPill.swift
import SwiftUI

// A port of Android's WavyPill.kt (cover/.../cover/components), which WavyPillTest covers there and
// WavyPillTests here.

/// Wave shape and speed for `WavyPill`, matching `PreviewWaveform`'s leading line.
let wavyPillWavelength: CGFloat = 40
let wavyPillWaveSpeed: CGFloat = 40

/// How far `WavyPill`'s edges swing at full amplitude. The pill is inset by this much.
let wavyPillMaxAmplitude: CGFloat = 3

/// Distance between the points a wavy edge is sampled at, in points.
private let sampleStep: CGFloat = 2

/// How far a wavy edge is displaced at `x`, for a straight run from `start` to `end` (the pill's
/// flat stretch between its round caps). A sine of `wavelength`, shifted by `phase`, scaled by
/// `amplitude`, and tapered to 0 over half a wavelength at either end so the wave meets the caps
/// without a kink. 0 outside the run.
func wavyEdgeOffset(x: CGFloat, start: CGFloat, end: CGFloat, amplitude: CGFloat, wavelength: CGFloat, phase: CGFloat) -> CGFloat {
    if x <= start || x >= end || amplitude == 0 { return 0 }
    let taper = min(wavelength / 2, (end - start) / 2)
    let fromEdge = min(x - start, end - x)
    let envelope = fromEdge >= taper ? 1 : smoothstep(fromEdge / taper)
    return amplitude * envelope * sin(2 * .pi * (x - phase) / wavelength)
}

private func smoothstep(_ t: CGFloat) -> CGFloat { t * t * (3 - 2 * t) }

/// A pill whose top and bottom edges are travelling waves: M3 Expressive's wavy look, as Android's
/// `wavyPillBackground` draws it. It's inset top and bottom by `maxAmplitude` so the waves never
/// leave its frame; its ends stay round.
///
/// `amplitude` is a 0...1 fraction of `maxAmplitude` (0 draws a plain pill) and animates; `phase` is
/// how far the wave has travelled, in points.
struct WavyPill: Shape {
    var amplitude: CGFloat
    var phase: CGFloat
    var wavelength: CGFloat = wavyPillWavelength
    var maxAmplitude: CGFloat = wavyPillMaxAmplitude

    var animatableData: CGFloat {
        get { amplitude }
        set { amplitude = newValue }
    }

    func path(in rect: CGRect) -> Path {
        let inset = maxAmplitude
        let top = rect.minY + inset
        let bottom = rect.maxY - inset
        let radius = max((bottom - top) / 2, 0)
        let start = rect.minX + radius
        let end = max(rect.maxX - radius, start)
        let swing = min(max(amplitude, 0), 1) * inset

        func offset(_ x: CGFloat) -> CGFloat {
            wavyEdgeOffset(x: x - rect.minX, start: radius, end: end - rect.minX, amplitude: swing, wavelength: wavelength, phase: phase)
        }

        // Top edge, left to right, then the right cap; bottom edge back, then the left cap. The
        // bottom mirrors the top, so the pill swells and pinches rather than bending.
        var path = Path()
        path.move(to: CGPoint(x: start, y: top))
        var x = start
        while x < end {
            x = min(x + sampleStep, end)
            path.addLine(to: CGPoint(x: x, y: top - offset(x)))
        }
        path.addArc(center: CGPoint(x: end, y: top + radius), radius: radius,
                    startAngle: .degrees(-90), endAngle: .degrees(90), clockwise: false)
        while x > start {
            x = max(x - sampleStep, start)
            path.addLine(to: CGPoint(x: x, y: bottom + offset(x)))
        }
        path.addArc(center: CGPoint(x: start, y: top + radius), radius: radius,
                    startAngle: .degrees(90), endAngle: .degrees(270), clockwise: false)
        path.closeSubpath()
        return path
    }
}

/// Drives a wave for `content`: an amplitude that eases between 0 and 1 as `running` changes, and a
/// clock, in seconds, for its phase. The clock runs while `running`, and keeps going while the
/// amplitude eases back to 0, so the wave settles flat while moving instead of freezing
/// mid-ripple. Idle, it asks for no frames. With Reduce Motion on, the clock stands still.
struct WaveAnimation<Content: View>: View {
    let running: Bool
    var animation: Animation = .spring(duration: 0.5, bounce: 0.2)
    @ViewBuilder let content: (_ amplitude: CGFloat, _ time: Double) -> Content

    @State private var amplitude: CGFloat = 0
    @State private var settling = false
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    var body: some View {
        TimelineView(.animation(paused: reduceMotion || (!running && !settling))) { timeline in
            content(amplitude, timeline.date.timeIntervalSinceReferenceDate)
        }
        .onAppear { amplitude = running ? 1 : 0 }
        .onChange(of: running) { _, running in
            settling = !running
            withAnimation(animation, completionCriteria: .logicallyComplete) {
                amplitude = running ? 1 : 0
            } completion: {
                settling = false
            }
        }
    }
}

/// How far a wave travelling at `speed` points per second has moved by `time`, within one `wavelength`.
func wavePhase(time: Double, speed: CGFloat, wavelength: CGFloat) -> CGFloat {
    CGFloat((time * Double(speed)).truncatingRemainder(dividingBy: Double(wavelength)))
}
